libyul/AsmParser.cpp

```diff
@@ -130,37 +130,73 @@ void Parser::fetchSourceLocationFromComment()
 	if (m_scanner->currentCommentLiteral().empty())
 		return;
 
-	static regex const lineRE = std::regex(
-		R"~~~((^|\s*)@src\s+(-1|\d+):(-1|\d+):(-1|\d+)(\s+|$))~~~",
-		std::regex_constants::ECMAScript | std::regex_constants::optimize
+	static regex const tagRegex = regex(
+		R"~~(\s*(@[a-zA-Z0-9\-_]+)\s+)~~",             // tag, e.g: @src
+		regex_constants::ECMAScript | regex_constants::optimize
+	);
+	static regex const srcTagArgsRegex = regex(
+		R"~~(^(-1|\d+):(-1|\d+):(-1|\d+)(?:\s+|$))~~", // index and location, e.g.: 1:234:-1
+		regex_constants::ECMAScript | regex_constants::optimize
 	);
 
-	string const text = m_scanner->currentCommentLiteral();
-	auto from = sregex_iterator(text.begin(), text.end(), lineRE);
-	auto to = sregex_iterator();
+	string const commentLiteral = m_scanner->currentCommentLiteral();
+	SourceLocation const commentLocation = m_scanner->currentCommentLocation();
+	smatch tagMatch;
+	string::const_iterator position = commentLiteral.begin();
 
-	for (auto const& matchResult: ranges::make_subrange(from, to))
+	while (regex_search(position, commentLiteral.end(), tagMatch, tagRegex))
 	{
-		solAssert(matchResult.size() == 6, "");
-
-		auto const sourceIndex = toInt(matchResult[2].str());
-		auto const start = toInt(matchResult[3].str());
-		auto const end = toInt(matchResult[4].str());
-
-		auto const commentLocation = m_scanner->currentCommentLocation();
-		m_debugDataOverride = DebugData::create();
-		if (!sourceIndex || !start || !end)
-			m_errorReporter.syntaxError(6367_error, commentLocation, "Invalid value in source location mapping. Could not parse location specification.");
-		else if (sourceIndex == -1)
-			m_debugDataOverride = DebugData::create(SourceLocation{*start, *end, nullptr});
-		else if (!(sourceIndex >= 0 && m_sourceNames->count(static_cast<unsigned>(*sourceIndex))))
-			m_errorReporter.syntaxError(2674_error, commentLocation, "Invalid source mapping. Source index not defined via @use-src.");
-		else
+		solAssert(tagMatch.size() == 2, "");
+		position += tagMatch.position() + tagMatch.length();
+
+		if (tagMatch[1] == "@src")
 		{
-			shared_ptr<string const> sourceName = m_sourceNames->at(static_cast<unsigned>(*sourceIndex));
-			solAssert(sourceName, "");
-			m_debugDataOverride = DebugData::create(SourceLocation{*start, *end, move(sourceName)});
+			smatch srcTagArgsMatch;
+			if (!regex_search(position, commentLiteral.end(), srcTagArgsMatch, srcTagArgsRegex))
```

Comment:
The only thing stopping us from using `Scanner` for parsing tag args is that there's no way to check where the scanner has stopped and use that to update `position`.

Reply:
But maybe the `^` already does it...

Reply:
Yeah, both would work (`regex_match()` with `.*` added at the end). It starts with `^` exactly for this purpose.